### PR TITLE
Graphics refactor #5

### PR DESCRIPTION
**Pull request by kkaefer**
Moves ProgramMap behind the `gl::Program` façade.

### CODE REVIEW

File: src/mbgl/gl/debugging.cpp [hunk 1]
```diff
@@ -9,8 +9,8 @@ using namespace platform;
 
 #ifndef NDEBUG
 
-DebugGroup::DebugGroup(const Context& context_, const std::string& name) : context(context_) {
-    if (auto debugging = context.getDebuggingExtension()) {
+DebugGroup::DebugGroup(const gfx::Context& context_, const std::string& name) : context(context_) {
+    if (auto debugging = reinterpret_cast<const gl::Context&>(context).getDebuggingExtension()) {
```
Comment (pozdnyakov) [inline src/mbgl/gl/debugging.cpp]: why not `static_cast`?

File: src/mbgl/util/literal.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace detail {
+
+using string_fn = const char* (*)();
+
+template <string_fn name, std::size_t... Is>
+string_literal<name()[Is]...> to_string_literal(std::index_sequence<Is...>);
+
+constexpr std::size_t string_length(const char* const str, std::size_t len = 0) {
+    return str[0] ? string_length(str + 1, len + 1) : len;
```
Comment (pozdnyakov) [inline src/mbgl/util/literal.hpp]: nit: `++str, ++len` ?

File: src/mbgl/util/literal.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+struct concat_literals<string_literal<Cs...>> {
+    static constexpr auto value() {
+        return string_literal<Cs...>::value();
+    }
+};
+
+} // namespace detail
+
+template <detail::string_fn... str>
+struct concat_literals {
```
Comment (pozdnyakov) [inline src/mbgl/util/literal.hpp]: great :-)

File: src/mbgl/gl/program.hpp [hunk 4]
```diff
@@ -31,93 +32,128 @@ class Program final : public gfx::Program<Name> {
...
-        const std::string fragmentSource = programs::gl::fragmentSource(programParameters, fragmentSource_);
+    const ProgramParameters programParameters;
+
+    static constexpr const auto vertexOffset = programs::gl::ShaderSource<Name>::vertexOffset;
+    static constexpr const auto fragmentOffset = programs::gl::ShaderSource<Name>::fragmentOffset;
+
+    class Instance {
+    public:
+        Instance(Context& context,
+                 const std::initializer_list<const char*>& vertexSource,
```
Comment (pozdnyakov) [inline src/mbgl/gl/program.hpp]: AFAIK `std::initializer_list` could be passed just by value

File: src/mbgl/gl/attribute.hpp [hunk 5]
```diff
@@ -95,5 +95,36 @@ class AttributeLocations<TypeList<As...>> final {
...
+    return "#define HAS_UNIFORM_u_";
+}
+
+template <class... As>
+class AttributeKey<TypeList<As...>> final {
+public:
+    static_assert(sizeof...(As) <= 32, "attribute count exceeds 32");
+
+    static uint32_t compute(const gfx::AttributeBindings<TypeList<As...>>& bindings) {
+        uint32_t value = 0;
```
Comment (pozdnyakov) [inline src/mbgl/gl/attribute.hpp]: nit: 0u

File: src/mbgl/util/literal.hpp (hunk 2 above)
Comment (kkaefer) [reply src/mbgl/util/literal.hpp]: why?

File: src/mbgl/gl/attribute.hpp (hunk 5 above)
Comment (kkaefer) [reply src/mbgl/gl/attribute.hpp]: why?

File: src/mbgl/util/literal.hpp (hunk 2 above)
Comment (kkaefer) [reply src/mbgl/util/literal.hpp]: also, the pointer is const, so we can't really call `++str`.

Comment (pozdnyakov) [reply src/mbgl/util/literal.hpp]: just more common way to represent iteration, I'm not insisting though - pls consider it as an optional nit :-) why does `str` need to be const?

**Comment by kkaefer** [comment]
will address `reinterpret_cast` -> `static_cast` in the next PR

File: src/mbgl/gl/attribute.hpp (hunk 5 above)
Comment (pozdnyakov) [reply src/mbgl/gl/attribute.hpp]: `0` is `int`, not `unsigned`.. however compiler optimizes it out anyway, so this point is optional too

**Pull request merged**